core/src/com/unciv/ui/mapeditor/TileEditorOptionsTable.kt

```diff
@@ -373,34 +375,60 @@ class TileEditorOptionsTable(val mapEditorScreen: MapEditorScreen): Table(Camera
-                if (tileInfo.getBaseTerrain().impassable) tileInfo.improvement = null
-                if (improvement.terrainsCanBeBuiltOn.isEmpty() && tileInfo.isWater)
-                    tileInfo.improvement = null
-                if (improvement.terrainsCanBeBuiltOn.isNotEmpty() // for "everywhere" improvements like city ruins, encampments, ancient ruins
-                        && improvement.terrainsCanBeBuiltOn.none { it == tileInfo.baseTerrain || it == tileInfo.terrainFeature })
-                    tileInfo.improvement = null
-            }
+        if (tileInfo.improvement!=null) {
+            normalizeTileImprovement(tileInfo)
         }
-        if(tileInfo.getBaseTerrain().impassable || tileInfo.isWater)
+        if (tileInfo.getBaseTerrain().impassable || tileInfo.isWater)
             tileInfo.roadStatus= RoadStatus.None
     }
 
+    private fun normalizeTileImprovement(tileInfo: TileInfo) {
+        if (tileInfo.improvement!!.startsWith("StartingLocation")) {
+            if (!tileInfo.isLand || tileInfo.getBaseTerrain().impassable)
+                tileInfo.improvement = null
+            return
+        }
+        val improvement = tileInfo.getTileImprovement()!!
+        val topTerrain = tileInfo.getLastTerrain()
+        val resource = if (tileInfo.resource!=null) tileInfo.getTileResource() else null
+        when {
+            // Precedence, simplified: terrainsCanBeBuiltOn, then improves-resource, then 'everywhere', default to false
+            // 'everywhere' improvements (city ruins, encampments, ancient ruins, great improvements)
+            // are recognized as without terrainsCanBeBuiltOn and with turnsToBuild == 0
+            "Cannot be built on bonus resource" in improvement.uniques
+                        && tileInfo.resource != null
+                        && resource?.resourceType == ResourceType.Bonus
+                -> tileInfo.improvement = null      // forbid if this unique matches
+            tileInfo.isLand      // Fishing boats have Coast allowed even though they're meant resource-only
+                        && topTerrain.name in improvement.terrainsCanBeBuiltOn
+                -> Unit     // allow where top terrain explicitly named
+            resource?.improvement == improvement.name
+                        && (!topTerrain.unbuildable || topTerrain.name in improvement.resourceTerrainAllow)
+                -> Unit     // allow where it improves a resource and feature OK
+            tileInfo.isWater || topTerrain.impassable
+                -> tileInfo.improvement = null      // forbid if water or mountains
+            improvement.terrainsCanBeBuiltOn.isEmpty() && improvement.turnsToBuild == 0
+                    // Allow Great Improvement but clear unbuildable terrain feature
+                    // Allow barbarian camps, ruins and similar without clear
+                -> if (topTerrain.unbuildable && improvement.name in Constants.greatImprovements)
```

Comment:
This is a nice catch, but should be without the `topTerrain.unbuildable` so great improvements will remove also forests and jungles

Reply:
...and they shouldn't remove features declared buildable. Tested.

By the way, I thought about not requiring that hardcoded list - and right now a !Stats().equals(improvement) || improvement.uniques.IsNotEmpty() would do it, too (at that point) - but nah, actually the json should make these tests obvious. What's a Great, what rules govern ruins etc and why has terrainsCanBeBuiltOn a different meaning on water than on land. To think about some other time.

Reply:
We do have a uniques list, we can have 'Great Improvement' be a unique.